File: minirys_ros2/include/minirys_ros2/nodes/pid.cpp
//#include "pid.hpp"
class PID
{
    float K;
    float Ti;
    float Td;
    float T;
    float e_past = 0.0f;
    float u_past = 0.0f;

public:
    PID(float T, float K=0.0f, float Ti=10000000.0f, float Td=0.0f){
        this->T = T;
        this->K = K;
        this->Td = Td;
        this->Ti = Ti;
    }

    float pid(float y, float y_zad){
        float e = y_zad - y;
        float up = this->K*e;
        float ui = this->u_past + this->K/this->Ti*this->T*(this->e_past + e)/2;
        float ud = this->K*this->Td*(e-this->e_past)/this->T;
        //float u = this->K*e + this->u_past + this->K/this->Ti*this->T*(this->e_past + e)/2 + this->K*this->Td*(e-this->e_past)/this->T;
        float u = up + ui + ud;
        this->u_past = ui;
        this->e_past = e;
        return u;
    }
    float p(float y, float y_zad){
        float e = y_zad - y;
        return this->K * e;
    }

    float pi(float y, float y_zad){
        float e = y_zad - y;
        float up = this->K*e;
        float ui = this->u_past + this->K/this->Ti*this->T*(this->e_past + e)/2;
        float u = up + ui;
        this->u_past = ui;
        this->e_past = e;
        return u;

    }
    float pd(float y, float y_zad){
        float e = y_zad - y;
        float up = this->K*e;
        float ud = this->K*this->Td*(e-this->e_past)/this->T;
        float u = up + ud;
        this->e_past = e;
        return u;

    }
    void clear(){
        this->e_past = 0.0f;
        this->u_past = 0.0f;
    }
};

File: followers/followers/pid.py
#!/usr/bin/env python3

class PID:
    K = None
    Ti = None
    Td = None
    T = None
    e_past = 0
    u_past = 0

    def __init__(self,T=0.5,K=0,Ti=100000000, Td= 0):
        self.K = K
        self.Ti = Ti
        self.Td = Td
        self.T = T
    
    def pid(self,y, y_zad):
        e = y_zad - y
        u = self.K*e + self.u_past + self.K/self.Ti*self.T*(self.e_past + e)/2 + self.K*self.Td*(e-self.e_past)/self.T
        self.u_past = self.u_past + self.K/self.Ti*self.T*(self.e_past + e)/2
        self.e_past = e
        return u

    def p(self,y, y_zad):
        e = y_zad - y
        u = self.K*e
        return u

    def pi(self,y, y_zad):
        e = y_zad - y
        u = self.K*e + self.u_past + self.K/self.Ti*self.T*(self.e_past + e)/2
        self.u_past = self.u_past + self.K/self.Ti*self.T*(self.e_past + e)/2
        self.e_past = e
        return u
    
    def pd(self,y, y_zad):
        e = y_zad - y
        u = self.K*e + self.K*self.Td*(e-self.e_past)/self.T
        self.e_past = e
        return u

    
